File: app/src/main/java/com/qingshuige/tangyuan/viewmodels/UserCardAdapter.java
package com.qingshuige.tangyuan.viewmodels;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;
import com.qingshuige.tangyuan.R;
import com.qingshuige.tangyuan.data.CircleTransform;
import com.qingshuige.tangyuan.network.ApiHelper;
import com.qingshuige.tangyuan.network.User;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class UserCardAdapter extends RecyclerView.Adapter<UserCardAdapter.ViewHolder> {
    List<UserInfo> userInfos;
    Context context;

    @NonNull
    @Override
    public UserCardAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_user_card, parent, false);
        return new UserCardAdapter.ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull UserCardAdapter.ViewHolder holder, int position) {
        UserInfo ui = userInfos.get(position);

        //Avatar
        Picasso.get()
                .load(ApiHelper.getFullImageURL(ui.getUser().avatarGuid))
                .resize(200, 0)
                .transform(new CircleTransform())
                .into(holder.getImgAvatar());

        //Nickname
        holder.getTextNickname().setText(ui.getUser().nickName);

        //isFollowed
        holder.getBtnFollow().setText(context.getString(ui.isFollowed() ? R.string.followed : R.string.follow));
    }

    @Override
    public int getItemCount() {
        return userInfos.size();
    }

    public void replaceDataset(List<UserInfo> list) {
        userInfos = list;
        notifyDataSetChanged();
    }

    public UserCardAdapter(Context context) {
        userInfos = new ArrayList<>();
        this.context = context;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private ImageView imgAvatar;
        private TextView textNickname;
        private MaterialButton btnFollow;

        public ViewHolder(@NonNull View view) {
            super(view);

            imgAvatar = view.findViewById(R.id.imgAvatar);
            textNickname = view.findViewById(R.id.textNickname);
            btnFollow = view.findViewById(R.id.btnFollow);
        }

        public ImageView getImgAvatar() {
            return imgAvatar;
        }

        public TextView getTextNickname() {
            return textNickname;
        }

        public MaterialButton getBtnFollow() {
            return btnFollow;
        }
    }
}

File: app/src/main/java/com/qingshuige/tangyuan/viewmodels/UserCardAdapter.kt
package com.qingshuige.tangyuan.viewmodels

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.button.MaterialButton
import com.qingshuige.tangyuan.R
import com.qingshuige.tangyuan.data.CircleTransform
import com.qingshuige.tangyuan.network.ApiHelper
import com.squareup.picasso.Picasso

class UserCardAdapter(private val context: Context) :
    RecyclerView.Adapter<UserCardAdapter.ViewHolder>() {

    private var userInfos = mutableListOf<UserInfo>()
    private var onUserClickListener: ItemActionListener? = null
    private var onFollowButtonClickListener: ItemActionListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view =
            LayoutInflater.from(parent.context).inflate(R.layout.view_user_card, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val userInfo = userInfos[position]

        // 头像加载
        Picasso.get()
            .load(ApiHelper.getFullImageURL(userInfo.user.avatarGuid))
            .resize(200, 0)
            .transform(CircleTransform())
            .into(holder.imgAvatar)

        // 昵称
        holder.textNickname.text = userInfo.user.nickName

        // 关注状态
        val followTextRes = if (userInfo.isFollowed) R.string.followed else R.string.follow
        holder.btnFollow.setText(followTextRes)

        // 点击事件
        holder.root.setOnClickListener {
            onUserClickListener?.onAction(userInfo)
        }

        holder.btnFollow.setOnClickListener {
            onFollowButtonClickListener?.onAction(userInfo)
        }
    }

    override fun getItemCount(): Int = userInfos.size

    fun replaceDataset(list: List<UserInfo>) {
        userInfos = list.toMutableList()
        notifyDataSetChanged()
    }

    fun setOnUserClickListener(listener: ItemActionListener) {
        this.onUserClickListener = listener
    }

    fun setOnFollowButtonClickListener(listener: ItemActionListener) {
        this.onFollowButtonClickListener = listener
    }

    fun interface ItemActionListener {
        fun onAction(info: UserInfo)
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val imgAvatar: ImageView = itemView.findViewById(R.id.imgAvatar)
        val textNickname: TextView = itemView.findViewById(R.id.textNickname)
        val btnFollow: MaterialButton = itemView.findViewById(R.id.btnFollow)
        val root: LinearLayout = itemView.findViewById(R.id.lnlUserCardRoot)
    }
}